tiatoolbox/models/engine/deep_feature_extractor.py

```diff
@@ -0,0 +1,557 @@
+            coordinates_zarr.shape[0] + coordinates_computed.shape[0],
+            *coordinates_zarr.shape[1:],
+        )
+    )
+    coordinates_zarr[-coordinates_computed.shape[0] :] = coordinates_computed
+
+    return probabilities_zarr, coordinates_zarr
+
+
+class DeepFeatureExtractor(SemanticSegmentor):
+    r"""Generic CNN-based feature extractor for digital pathology images.
+
+    This class extends :class:`SemanticSegmentor` to extract deep features from
+    whole slide images (WSIs) or image patches using a CNN model. It is designed
+    for use cases where the goal is to obtain intermediate feature representations
+    (e.g., embeddings) rather than final classification or segmentation outputs.
+
+    The extracted features are returned or saved in Zarr format for downstream
+    analysis, such as clustering, visualization, or training other machine learning
+    models.
+
+    Args:
+        model (str | ModelABC):
+            A PyTorch model instance or the name of a pretrained model from TIAToolbox.
+        batch_size (int):
+            Number of image patches processed per forward pass. Default is 8.
+        num_workers (int):
+            Number of workers for data loading. Default is 0.
+        weights (str | Path | None):
+            Path to model weights. If None, default weights are used.
+        device (str):
+            Device to run the model on (e.g., "cpu", "cuda"). Default is "cpu".
+        verbose (bool):
+            Whether to enable verbose logging. Default is True.
+
+    Attributes:
+        process_prediction_per_batch (bool):
+            Flag to control whether predictions are processed per batch.
+            Default is False.
+
+    """
+
+    def __init__(
+        self: DeepFeatureExtractor,
+        model: str | ModelABC,
+        batch_size: int = 8,
+        num_workers: int = 0,
+        weights: str | Path | None = None,
+        *,
+        device: str = "cpu",
+        verbose: bool = True,
+    ) -> None:
+        """Initialize :class:`DeepFeatureExtractor`.
+
+        Args:
+            model (str | ModelABC):
+                A PyTorch model instance or the name of a pretrained model from
+                TIAToolbox. If a string is provided, the corresponding pretrained
+                weights will be downloaded unless overridden via `weights`.
+            batch_size (int):
+                Number of image patches processed per forward pass. Default is 8.
+            num_workers (int):
+                Number of workers for data loading. Default is 0.
+            weights (str | Path | None):
+                Path to model weights. If None, default weights are used.
+            device (str):
+                Device to run the model on (e.g., "cpu", "cuda"). Default is "cpu".
+            verbose (bool):
+                Whether to enable verbose logging. Default is True.
+
+        """
+        super().__init__(
+            model=model,
+            batch_size=batch_size,
```

Comment:
when a string is passed as a model, it should really go look in the lists of architectures available for TIMMBackbone and CNNBackbone (in architecture.vanilla.py) and build the appropriate model. At the moment it looks in the models defined in pretrainedModel.yaml which is patch prediction/segmentation models. And those don't work in here.

We may even want to consolidate these, so that there is a single unified way of getting models, whatever type they are.

Reply:
Ideally, the feature extractor should work with both `TimmBackbone("efficientnet_b0", pretrained=True)` and `"efficientnet_b0"` options.